scripts/rddepman.ts

```diff
@@ -0,0 +1,210 @@
+  const currentVersions = await readDependencyVersions(depVersionsPath);
+
+  // get a list of dependencies' version comparisons
+  const versionComparisons: VersionComparison[] = await Promise.all(dependencies.map(async(dependency) => {
+    return {
+      name:           dependency.name,
+      currentVersion: currentVersions[dependency.name as keyof DependencyVersions],
+      latestVersion:  await dependency.getLatestVersion(),
+    };
+  }));
+
+  // filter comparisons down to the ones that need an update
+  const updatesAvailable = versionComparisons.filter(({ name, currentVersion, latestVersion }) => {
+    const equal = compareVersions(currentVersion, latestVersion);
+
+    if (equal) {
+      console.log(`${ name } is up to date.`);
+    } else {
+      console.log(`Can update ${ name } from ${ currentVersion } to ${ latestVersion }`);
+    }
+
+    return !equal;
+  });
+
+  // reconcile dependencies that need an update with state of repo's PRs
+  const needToCreatePR: VersionComparison[] = [];
+
+  await Promise.all(updatesAvailable.map(async({ name, currentVersion, latestVersion }) => {
+    // try to find PR for this combo of name, current version and latest version
+    const prs = await getPulls(name);
+
+    // we use title, rather than branch name, to filter pull requests
+    // because branch name is not available from the search endpoint
+    const title = getTitle(name, currentVersion, latestVersion);
+    let prExists = false;
+
+    await Promise.all(prs.map(async(pr) => {
+      if (pr.title !== title && pr.state === 'open') {
+        console.log(`Closing stale PR "${ pr.title }".`);
+        await getOctokit().rest.pulls.update({
+          owner: GITHUB_OWNER, repo: GITHUB_REPO, pull_number: pr.number, state: 'closed',
+        });
+      } else if (pr.title === title) {
+        console.log(`Found existing PR "${ title }".`);
+        prExists = true;
+      }
+    }));
+    if (!prExists) {
+      console.log(`Could not find PR "${ title }". Will create.`);
+      needToCreatePR.push({
+        name, currentVersion, latestVersion,
+      });
+    }
+  }));
+
+  // create a branch for each version update, make changes, and make a PR from the branch
+  for (const { name, currentVersion, latestVersion } of needToCreatePR) {
+    const branchName = getBranchName(name, currentVersion, latestVersion);
+    const commitMessage = `Bump ${ name } from ${ currentVersion } to ${ latestVersion }`;
+
+    git('checkout', '-b', branchName, MAIN_BRANCH);
+    const depVersions = await readDependencyVersions(depVersionsPath);
+
+    depVersions[name as keyof DependencyVersions] = latestVersion as string & AlpineLimaISOVersion;
+    await writeDependencyVersions(depVersionsPath, depVersions);
+    git('add', depVersionsPath);
+    git('commit', '--signoff', '--message', commitMessage);
+    git('push', '--force', `https://${ process.env.GITHUB_TOKEN }@github.com/${ GITHUB_OWNER }/${ GITHUB_REPO }`);
```

Comment:
I believe (because we checked out with permissions) we don't need to set the remote here, just `git push --force origin ${branch}`?

Reply:
Because it is possible to specify the owner (i.e. rancher-sandbox, or mook-as, or adamkpickering) of the repository via `GITHUB_REPOSITORY_OWNER`, we need to be explicit about where we're pushing the branch. I ran into this yesterday while testing - I accidentally pushed a branch to rancher-sandbox/rancher-desktop rather than adamkpickering/rancher-desktop.

Reply:
According to https://docs.github.com/en/actions/security-guides/automatic-token-authentication:

> Before each job begins, GitHub fetches an installation access token for the job. The GITHUB_TOKEN expires when a job finishes or after a maximum of 24 hours.

A clarification: https://stackoverflow.com/questions/73643738/secrets-github-token-could-be-expired-in-github-action

It seems that the token will expire when the job finishes, or when 24 hours has elapsed, whenever comes first. So unless the command is printed while the workflow is running I think we are safe. In the event of a failure, the token will expire anyways. Let me know what you think.